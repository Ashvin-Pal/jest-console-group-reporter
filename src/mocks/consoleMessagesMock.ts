import type { ConsoleMessage } from "@/types";

export const consoleMessagesMock: ConsoleMessage[] = [
  {
    type: "error",
    message: "Network connection lost.",
    origin: "Network Module",
  },
  {
    type: "error",
    message: "Network connection lost.",
    origin: "Network Module",
  },
  {
    type: "error",
    message: "There was an error processing the data.",
    origin: "Server",
  },
  {
    type: "warn",
    message: "Server timeout.",
    origin: "Server",
  },
  {
    type: "warn",
    message: "Resource not found.",
    origin: "Resource Loader",
  },
  {
    type: "info",
    message: "User profile updated.",
    origin: "User Profile",
  },
  {
    type: "info",
    message:
      "This is an exceptionally lengthy info message. It ensures that extremely long messages are truncated when they exceed 150 characters in test summaries.",
    origin: "User Profile",
  },
  {
    type: "log",
    message: "This is a sample log message.",
    origin: "App",
  },
];
