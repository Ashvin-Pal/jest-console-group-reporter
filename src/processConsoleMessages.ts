import type { TestResult, Config } from "@jest/reporters";
import type { ConsoleMessagesMap, Options } from "@/types";
import { getGroupName } from "./getGroupName";
import { shouldFilterMessage } from "./shouldFilterMessage";
import { mergeLogEntries, createLogEntry } from "@/logEntry";

interface ProcessConsoleMessagesOptions
  extends Pick<Options, "filters" | "groups">,
    Pick<TestResult, "testFilePath" | "console"> {
  config: Config.ProjectConfig;
}

export function processConsoleMessages({
  filters,
  groups,
  testFilePath,
  console: consoleMessages = [],
  config,
}: ProcessConsoleMessagesOptions): {
  consoleMessagesMap: ConsoleMessagesMap;
  filteredCount: number;
} {
  const consoleMessagesMap: ConsoleMessagesMap = new Map();
  let filteredCount = 0;

  for (const consoleMessage of consoleMessages) {
    if (shouldFilterMessage(filters, consoleMessage)) {
      filteredCount += 1;
      continue;
    }

    const { groupName, isCustomGroup } = getGroupName(consoleMessage, groups);

    let logEntryMap = consoleMessagesMap.get(consoleMessage.type);
    if (!logEntryMap) {
      logEntryMap = new Map();
      consoleMessagesMap.set(consoleMessage.type, logEntryMap);
    }

    const logEntry = createLogEntry({ isCustomGroup, testFilePath, ...consoleMessage, config });

    const existingEntry = logEntryMap.get(groupName);
    logEntryMap.set(groupName, existingEntry ? mergeLogEntries(existingEntry, logEntry) : logEntry);
  }

  return { consoleMessagesMap, filteredCount };
}
