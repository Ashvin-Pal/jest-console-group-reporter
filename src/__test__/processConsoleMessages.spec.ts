import type { Config } from "@jest/reporters";
import { processConsoleMessages } from "../processConsoleMessages";
import { consoleMessagesMock } from "../mocks";

const testFilePath = "test.js";
const config = { projectRoot: "./" } as unknown as Config.ProjectConfig;

describe("processConsoleMessages", () => {
  test("Should return console messages map with no custom groups or filters", () => {
    const result = processConsoleMessages({
      filters: [],
      groups: [],
      testFilePath,
      // @ts-expect-error Issues with jest type
      console: consoleMessagesMock,
      config,
    });

    expect(result.filteredCount).toBe(0);
    expect(result.consoleMessagesMap).toMatchSnapshot();
  });

  test("Should filter all console messages", () => {
    const result = processConsoleMessages({
      filters: [(): boolean => true],
      groups: [],
      testFilePath,
      // @ts-expect-error Issues with jest type
      console: consoleMessagesMock,
      config,
    });

    expect(result.consoleMessagesMap.size).toBe(0);
    expect(result.filteredCount).toBe(consoleMessagesMock.length);
  });

  test("Should group all console messages under one group name for each console type", () => {
    const groupName = "All";
    const result = processConsoleMessages({
      filters: [],
      groups: [
        {
          name: groupName,
          match: (): boolean => true,
        },
      ],
      testFilePath,
      // @ts-expect-error Issues with jest type
      console: consoleMessagesMock,
      config,
    });

    expect(result.consoleMessagesMap.size).toBe(4);
    expect(result.filteredCount).toBe(0);

    const consoleMessagesEntries = result.consoleMessagesMap.entries();

    for (const [, logEntryMap] of consoleMessagesEntries) {
      expect(logEntryMap.size).toBe(1);
      const logEntry = logEntryMap.get(groupName);
      expect(logEntry).toBeDefined();
      expect(logEntry?.isCustomGroup).toBe(true);
      expect(logEntry?.testFilePaths).toStrictEqual(new Set(["test.js"]));
      expect(logEntry?.messages).toBeInstanceOf(Set);
      expect(logEntry?.stackTrace).toBeDefined();
      expect(logEntry?.count).toBeGreaterThan(0);
    }
  });

  test("Should filter and group messages correctly", () => {
    const result = processConsoleMessages({
      filters: [/^This is a sample log message/],
      groups: [
        {
          name: "Error group",
          match: ({ type }): boolean => type === "error",
        },
      ],
      testFilePath,
      // @ts-expect-error Issues with jest type
      console: consoleMessagesMock,
      config,
    });

    const errorMap = result.consoleMessagesMap.get("error");
    const errorCustomLogEntryGroup = errorMap?.get("Error group");
    expect(errorMap?.size).toBe(1);
    expect(result.filteredCount).toBe(1);
    expect(errorCustomLogEntryGroup).toStrictEqual({
      count: 3,
      isCustomGroup: true,
      stackTrace: "Network Module",
      testFilePaths: new Set(["test.js"]),
      messages: new Set(["Network connection lost.", "There was an error processing the data."]),
    });
    expect(result.consoleMessagesMap.size).toBe(3);
    expect(result.consoleMessagesMap).toMatchSnapshot();
  });
});
