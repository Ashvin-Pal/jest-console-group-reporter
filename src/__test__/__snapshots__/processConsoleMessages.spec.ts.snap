// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processConsoleMessages Should filter and group messages correctly 1`] = `
Map {
  "error" => Map {
    "Error group" => {
      "count": 3,
      "isCustomGroup": true,
      "messages": Set {
        "Network connection lost.",
        "There was an error processing the data.",
      },
      "stackTrace": "Network Module",
      "testFilePaths": Set {
        "test.js",
      },
    },
  },
  "warn" => Map {
    "Server timeout." => {
      "count": 1,
      "isCustomGroup": false,
      "stackTrace": "Server",
      "testFilePaths": Set {
        "test.js",
      },
    },
    "Resource not found." => {
      "count": 1,
      "isCustomGroup": false,
      "stackTrace": "Resource Loader",
      "testFilePaths": Set {
        "test.js",
      },
    },
  },
  "info" => Map {
    "User profile updated." => {
      "count": 1,
      "isCustomGroup": false,
      "stackTrace": "User Profile",
      "testFilePaths": Set {
        "test.js",
      },
    },
    "This is an exceptionally lengthy info message. It ensures that extremely long messages are truncated when they exceed 150 characters in test summaries." => {
      "count": 1,
      "isCustomGroup": false,
      "stackTrace": "User Profile",
      "testFilePaths": Set {
        "test.js",
      },
    },
  },
}
`;

exports[`processConsoleMessages Should return console messages map with no custom groups or filters 1`] = `
Map {
  "error" => Map {
    "Network connection lost." => {
      "count": 2,
      "isCustomGroup": false,
      "stackTrace": "Network Module",
      "testFilePaths": Set {
        "test.js",
      },
    },
    "There was an error processing the data." => {
      "count": 1,
      "isCustomGroup": false,
      "stackTrace": "Server",
      "testFilePaths": Set {
        "test.js",
      },
    },
  },
  "warn" => Map {
    "Server timeout." => {
      "count": 1,
      "isCustomGroup": false,
      "stackTrace": "Server",
      "testFilePaths": Set {
        "test.js",
      },
    },
    "Resource not found." => {
      "count": 1,
      "isCustomGroup": false,
      "stackTrace": "Resource Loader",
      "testFilePaths": Set {
        "test.js",
      },
    },
  },
  "info" => Map {
    "User profile updated." => {
      "count": 1,
      "isCustomGroup": false,
      "stackTrace": "User Profile",
      "testFilePaths": Set {
        "test.js",
      },
    },
    "This is an exceptionally lengthy info message. It ensures that extremely long messages are truncated when they exceed 150 characters in test summaries." => {
      "count": 1,
      "isCustomGroup": false,
      "stackTrace": "User Profile",
      "testFilePaths": Set {
        "test.js",
      },
    },
  },
  "log" => Map {
    "This is a sample log message." => {
      "count": 1,
      "isCustomGroup": false,
      "stackTrace": "App",
      "testFilePaths": Set {
        "test.js",
      },
    },
  },
}
`;
